src/input/type_base.cc

```diff
@@ -109,6 +110,20 @@ void TypeBase::add_attribute(std::string name, json_string val) {
 
 
 void TypeBase::write_attributes(ostream& stream) const {
+	// print hash of generic type and parameters into separate keys
+	if (generic_type_.size()) {
```

Comment:
Output os generic_type_ and parameter_map_ do not match the name of method etc. Move as much as possible into TypeOutputBase. Any reason to have write_attributes is in TypeBase?